R/set-record-option.R

```diff
@@ -0,0 +1,110 @@
+
+#' Add or modify a record option
+#'
+#' Add a new record option or adjust the value/presence of an existing one.
+#'
+#' Note that only estimation, data, and table records are currently supported.
+#'
+#' Notes on `value`:
+#'  - We infer whether to use `option_value` or `option_flag` for *new* options based on the *class* of `value`:
+#'     - If you are trying to add an option with an assigned value (i.e. `MAXEVAL = 10`), `value` must be a character string or numeric value
+#'     - If you are trying to add a flag option (i.e. `NOABORT`), `value` should be logical (`T`/`F`).
+#'  - If `value` is set to `NULL`, the option will be removed (if found). `FALSE` is also acceptable for removing an existing option of class `nmrec_option_flag`.
+#'
+#' @param record An [nmrec_record] object.
+#' @param name Name of record to select. Any valid spelling of the option name
+#'   is allowed.
+#' @param value value to assign to option.
+#'
+#' @seealso `nmrec::option` selecting records of a given type.
+#'
+#'
+set_record_option <- function(record,
+                              name,
+                              value) {
+  stopifnot(inherits(record, "nmrec_record"))
+  stopifnot(length(name) == 1, is.character(name), nzchar(name))
+  stopifnot(record[["name"]] %in% c("estimation", "data", "table"))
+
+
+  name_map <- get(paste0(record[["name"]], "_option_names"))
+  name_resolved <- get0(tolower(name), name_map)
+  rec_opt_prev <- get_record_option(record, name)
+
+
+  if (is.null(name_resolved) && is.null(rec_opt_prev)) {
+    abort(
+      sprintf("%s is not a valid option for %s blocks", name, record[["name"]]),
+      nmrec_error()
+    )
+  }
+
+  # Create new record option
+  if (is.null(rec_opt_prev)) {
+    if (is.null(value)) {
+      abort(
+        sprintf("Option %s does not exist in this record. `NULL` is only valid for removing existing options.", name),
+        nmrec_error()
+      )
+    } else {
+      opt_type <- if (is.logical(value)) "flag" else "value"
+
+      if (opt_type == "value") {
+        stopifnot(is.numeric(value) || is.character(value))
+        opt <- option_value$new(name_resolved, name, value = value)
+      } else if (opt_type == "flag") {
+        stopifnot(is.logical(value))
+        opt <- option_flag$new(name_resolved, name)
+      }
+    }
+
+    new_opt_lst <- list(elem_whitespace(" "), opt)
```

Comment:
I'd prefer you put the new option on its own new line.  Aside from more aesthetic reasons, this has the piratical advantage of not needing to worry whether then end of the last line was a continuation marker or comment.

Reply:
I can do this, but honestly feel I should push back. I think aesthetic should very much be a consideration here, given that it's modelers, and not developers, that would go through the `ctl` files at times. If you think `find_record_end` is potentially risky, I can do some testing with more edge cases, but I think a more visually seamless amendment of the record would be preferable

Reply:
I find it aesthetically more appealing if programatic modifications are put on their own line, so it's very much eye of the beholder.  I already mentioned two practical issues that using a new line resolves.  Of course you could deal with those (and perhaps others I'm not thinking of), but I don't it's not worth the complication.  (And it can be adjusted later if we decide the using a new line is a eyesore that must be avoided.)

Reply:
Discussed this today on call with @barrettk.  Talked through two complications of not using a new line: comments and `&`-style continuations.  

He had the idea that finding that walking backward and finding the first option may be a good way to find the right position.  He's going to play around with that a bit.

Reply:
> He had the idea that finding that walking backward and finding the first option may be a good way to find the right position. He's going to play around with that a bit.

That direction worked out nicely.  Associated tests: https://github.com/metrumresearchgroup/nmrec/blob/e67cc6e9ba175c87184302ca5f98729115f03a4a/tests/testthat/test-set-record-option.R#L119-L207